### PR TITLE
Tooltip enhancement

### PR DESCRIPTION
**Pull request by ldong**
Arrow pointer enhancement
1. Add _up_ and _down_ arrow support
2. Add mask
3. Add auto-resize
4. Fix typo `initTootltip` to `initTooltip`

Screenshot:

tooltip

### CODE REVIEW

File: addon/components/ember-tutorial-component.js [hunk 1]
```diff
@@ -167,17 +179,37 @@ export default Ember.Component.extend({
...
+      this.setPosition();
+    }.bind(this));
+  },
+
+  /**
+   * Ember Hook used to evit event listeners
+   * @method didDestroyElement
+   */
+  didDestroyElement() {
+    console.log('resize position and trigger setPosition');
```
Comment (svkangal) [inline addon/components/ember-tutorial-component.js]: Remove console log. User Ember Logger instead

File: addon/components/ember-tutorial-component.js [hunk 2]
```diff
@@ -191,27 +223,120 @@ export default Ember.Component.extend({
...
     } else {
       this.set('hideMessage', true);
     }
   },
 
   /**
+   * @method turn off resize listeners
+   */
+  cancelAutoResize() {
+    console.log('cancelAutoResize');
```
Comment (svkangal) [inline addon/components/ember-tutorial-component.js]: Remove console log. User Ember Logger instead

File: addon/components/ember-tutorial-component.js [hunk 3]
```diff
@@ -191,27 +223,120 @@ export default Ember.Component.extend({
...
+    } else {
+      this.computeXCord();
+      this.computeYCord();
+    }
+  },
+
+  computeXYCord(tooltipElementSelector='.tutorial-component') {
+    let $mainElement = Ember.$(this.get('currentConfig').ele);
+    let $tooltipElement = Ember.$(tooltipElementSelector);
+    const borderWidth = this.getBorderWidth(tooltipElementSelector);
```
Comment (svkangal) [inline addon/components/ember-tutorial-component.js:302]: Do we need to calculate border width? getBoundingClientRect gives you width and height including the border widths and padding.

Comment (svkangal) [inline addon/components/ember-tutorial-component.js:302]: No worries I see now what it has been used for.

File: addon/components/ember-tutorial-component.js [hunk 4]
```diff
@@ -191,27 +223,120 @@ export default Ember.Component.extend({
...
+        break;
+      default:
+        Ember.Logger.error('This condition should not reach');
+    }
+
+    this.set('xCoord', x);
+    this.set('yCoord', y);
+  },
+
+  didInsertElement() {
```
Comment (svkangal) [inline addon/components/ember-tutorial-component.js:347]: why do we need this?

File: addon/templates/components/ember-tutorial-component.hbs [hunk 5]
```diff
@@ -1,5 +1,6 @@
 {{#ember-wormhole to="ember-tutorial-component-wormhole"}}
-  <div class='tutorial-component {{if isLeftPointer "left" "right"}} {{if hideMessage "hide" "show"}}' style={{tooltipStyle}}>
+<div class='tutorial-component-mask {{if hideMessage "hide" "show"}}'>
```
Comment (svkangal) [inline addon/templates/components/ember-tutorial-component.hbs:3]: Lets keep tutorial-component-mask optional and part of config.

File: addon/components/ember-tutorial-component.js [hunk 6]
```diff
@@ -167,17 +179,37 @@ export default Ember.Component.extend({
...
+
+  /**
+   * Overrides the init function
+   */
+  init() {
+    this._super(...arguments);
+
+    Ember.$(window).on('resize.tutorial-component', function() {
+      console.log('resize position and trigger setPosition');
+      this.setPosition();
```
Comment (svkangal) [inline addon/components/ember-tutorial-component.js:209]: Thanks for adding this, wanted to add that in the initial release itself but missed it.

File: addon/components/ember-tutorial-component.js [hunk 7]
```diff
@@ -191,27 +223,120 @@ export default Ember.Component.extend({
...
+    if (H >= TH) {
+      heightSign = 1;
+    } else {
+      heightSign = -1;
+    }
+
+    const position = this.get('pointerDirection').toLowerCase();
+    switch(position) {
+      case 'left':
+        x = X + W + borderWidth;
```
Comment (svkangal) [inline addon/components/ember-tutorial-component.js:324]: lets discuss this logic, have some questions.

File: addon/components/ember-tutorial-component.js [hunk 8]
```diff
@@ -191,27 +223,120 @@ export default Ember.Component.extend({
...
+  setPosition() {
+    if (this.get('currentConfig').pointerDirection) {
+      this.computeXYCord();
+    } else {
+      this.computeXCord();
+      this.computeYCord();
+    }
+  },
+
+  computeXYCord(tooltipElementSelector='.tutorial-component') {
```
Comment (svkangal) [inline addon/components/ember-tutorial-component.js:299]: Please add some documentation every where. Also if possible add some inline comments explaining the code.

**Comment by svkangal** [comment]
Will be great if you can add some tests too and also update the Readme.

**Comment by ldong** [comment]
1. Change console.log to Ember Log
2. Add documentation to indicate the usage of up, right, down, left for arrow pointer position
3. Add mask as optional field into config
4. Add mask button in the demo page, application.hbs
5. Update _README_

**Pull request merged**